### ISSUE TITLE
Different testing accuracy

### ISSUE THREAD

**Issue by choijenna64** [eriklindernoren/PyTorch-YOLOv3#119]
Running the model on test.py on the coco dataset consistently outputs around 40mAP instead of the 57.9 or 58.2. Is there a reason why I cannot get this 57.9 or 58.2mAP output? 
<img width="279" alt="screen shot 2019-02-18 at 2 36 35 pm" src="https://user-images.githubusercontent.com/36273841/52973015-b499ea00-338a-11e9-8635-fa926fb4b61b.png">

**Comment by fourth-archive** [comment]
The repo below tests to about 0.58 mAP on COCO using the original YOLOv3 weights:
https://github.com/ultralytics/yolov3

If you run `python3 test.py` you should see:
```
      Image      Total  Precision     Recall        mAP
       5000       5000      0.633      0.598      0.589

mAP Per Class:
         person: 0.7397
        bicycle: 0.4354
            car: 0.4884
      motorbike: 0.6372
      aeroplane: 0.8263
            bus: 0.7101
          train: 0.7713
          truck: 0.3599
           boat: 0.3982
  traffic light: 0.4359
   fire hydrant: 0.7410
      stop sign: 0.7251
  parking meter: 0.4293
          bench: 0.2846
           bird: 0.4764
            cat: 0.6460
            dog: 0.5972
          horse: 0.6855
          sheep: 0.4297
            cow: 0.4343
       elephant: 0.8016
           bear: 0.6418
          zebra: 0.7726
        giraffe: 0.8707
       backpack: 0.2034
       umbrella: 0.5101
        handbag: 0.1676
            tie: 0.5130
       suitcase: 0.3754
        frisbee: 0.6494
           skis: 0.4402
      snowboard: 0.5657
    sports ball: 0.5956
           kite: 0.5647
   baseball bat: 0.5436
 baseball glove: 0.5312
     skateboard: 0.7109
      surfboard: 0.6562
  tennis racket: 0.7707
         bottle: 0.3868
     wine glass: 0.4738
            cup: 0.4165
           fork: 0.3319
          knife: 0.2303
          spoon: 0.2031
           bowl: 0.3590
         banana: 0.3034
          apple: 0.1920
       sandwich: 0.3489
         orange: 0.2760
       broccoli: 0.3100
         carrot: 0.1926
        hot dog: 0.4404
          pizza: 0.5814
          donut: 0.4284
           cake: 0.4452
          chair: 0.3541
           sofa: 0.4362
    pottedplant: 0.3704
            bed: 0.5254
    diningtable: 0.3670
         toilet: 0.8059
      tvmonitor: 0.6290
         laptop: 0.6277
          mouse: 0.6213
         remote: 0.3764
       keyboard: 0.5638
     cell phone: 0.2963
      microwave: 0.5795
           oven: 0.4246
        toaster: 0.0000
           sink: 0.5452
   refrigerator: 0.5449
           book: 0.1520
          clock: 0.6236
           vase: 0.4339
       scissors: 0.2896
     teddy bear: 0.5438
     hair drier: 0.0000
     toothbrush: 0.2697
```

**Comment by RenAihua** [comment]
@choijenna64 hi, i met the same problem and i just got the 0.4096. what's more  i can't understand the class 70 and 78 got zero. have you got the idea.

**Comment by 0801130205** [comment]
@choijenna64 @RenAihua Hi, I met the same problem and I just got the mAP: 0.40963824871843535. Have you solved the problem?

**Comment by fourth-archive** [comment]
@RenAihua @choijenna64 use this repo. **It outputs official COCO mAP using pycocotools**.
https://github.com/ultralytics/yolov3

``` bash
sudo rm -rf yolov3 && git clone https://github.com/ultralytics/yolov3
# bash yolov3/data/get_coco_dataset.sh
sudo rm -rf cocoapi && git clone https://github.com/cocodataset/cocoapi && cd cocoapi/PythonAPI && make && cd ../.. && cp -r cocoapi/PythonAPI/pycocotools yolov3
cd yolov3 && python3 test.py
```

**Comment by liujiang137** [comment]
Hi, I met the same problem ，Now I am training my model and I want to know  the best result you got on this code . Thank you in advance for your reply.

**Comment by fourth-archive** [comment]
@liujiang137 from https://github.com/ultralytics/yolov3#map

# mAP

- Use `test.py --weights weights/yolov3.weights` to test the official YOLOv3 weights.
- Use `test.py --weights weights/latest.pt` to test the latest training results.
- Compare to official darknet results from https://arxiv.org/abs/1804.02767.

<i></i> | ultralytics/yolov3 | darknet  
--- | ---| ---   
YOLOv3-320 | 51.3 | 51.5  
YOLOv3-416 | 54.9 | 55.3  
YOLOv3-608 | 57.9 | 57.9  

``` bash
sudo rm -rf yolov3 && git clone https://github.com/ultralytics/yolov3
# bash yolov3/data/get_coco_dataset.sh
sudo rm -rf cocoapi && git clone https://github.com/cocodataset/cocoapi && cd cocoapi/PythonAPI && make && cd ../.. && cp -r cocoapi/PythonAPI/pycocotools yolov3
cd yolov3

python3 test.py --save-json --conf-thres 0.001 --img-size 416
Namespace(batch_size=32, cfg='cfg/yolov3.cfg', conf_thres=0.001, data_cfg='cfg/coco.data', img_size=416, iou_thres=0.5, nms_thres=0.45, save_json=True, weights='weights/yolov3.weights')
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.308
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.549
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.310
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.141
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.334
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.454
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.267
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.403
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.428
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.237
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.464
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.585
 
python3 test.py --save-json --conf-thres 0.001 --img-size 608 --batch-size 16
Namespace(batch_size=16, cfg='cfg/yolov3.cfg', conf_thres=0.001, data_cfg='cfg/coco.data', img_size=608, iou_thres=0.5, nms_thres=0.45, save_json=True, weights='weights/yolov3.weights')
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.328
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.579
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.335
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.190
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.357
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.428
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.279
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.429
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.456
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.299
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.483
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.572
```

**Comment by liujiang137** [comment]
> @liujiang137 from https://github.com/ultralytics/yolov3#map
> mAP
> 
> Use test.py --weights weights/yolov3.weights to test the official YOLOv3 weights.
> Use test.py --weights weights/latest.pt to test the latest training results.
> Compare to official darknet results from https://arxiv.org/abs/1804.02767.
> 
> 
> 
> 
> 
> ultralytics/yolov3
> darknet
> 
> 
> 
> 
> YOLOv3-320
> 51.3
> 51.5
> 
> 
> YOLOv3-416
> 54.9
> 55.3
> 
> 
> YOLOv3-608
> 57.9
> 57.9
> 
> 
> 
> sudo rm -rf yolov3 && git clone https://github.com/ultralytics/yolov3
> # bash yolov3/data/get_coco_dataset.sh
> sudo rm -rf cocoapi && git clone https://github.com/cocodataset/cocoapi && cd cocoapi/PythonAPI && make && cd ../.. && cp -r cocoapi/PythonAPI/pycocotools yolov3
> cd yolov3
> 
> python3 test.py --save-json --conf-thres 0.001 --img-size 416
> Namespace(batch_size=32, cfg='cfg/yolov3.cfg', conf_thres=0.001, data_cfg='cfg/coco.data', img_size=416, iou_thres=0.5, nms_thres=0.45, save_json=True, weights='weights/yolov3.weights')
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.308
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.549
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.310
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.141
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.334
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.454
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.267
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.403
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.428
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.237
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.464
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.585
>  
> python3 test.py --save-json --conf-thres 0.001 --img-size 608 --batch-size 16
> Namespace(batch_size=16, cfg='cfg/yolov3.cfg', conf_thres=0.001, data_cfg='cfg/coco.data', img_size=608, iou_thres=0.5, nms_thres=0.45, save_json=True, weights='weights/yolov3.weights')
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.328
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.579
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.335
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.190
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.357
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.428
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.279
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.429
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.456
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.299
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.483
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.572

Thank you very much for your reply. I read the requirements of this code and found that it needs to run on python3.7. My server is python 3.5, what is the incompatibility between Python 3.5 and 3.7? What do I need to modify?
